### PR TITLE
Add button for algorithms

### PR DESCRIPTION
**Pull request by pierreosselin**
- Latex description added at bottom page (can't change the font-size)
- New 1D function added
- Algorithm Classes better organized
- Momentum algorithm Added
- Button for managing algorithm added for graph and Contour (keeping the initial points for better comparison)

### CODE REVIEW

File: app.html [hunk 1]
```diff
@@ -5,10 +5,18 @@
   <title>Title</title>
   <link rel="stylesheet" type="text/css" href="./graph.css"  />
   <script src="https://d3js.org/d3.v5.min.js"></script>
+  <script src="latex.js"></script>
+  <!--<script type="text/javascript" src="http://latex.codecogs.com/latexit.js"></script>-->
+  <!--<script type="text/javascript"> // Can be deleted if latex not used with <p>
+  LatexIT.add('p',true);
+  </script> -->
```
Comment (Scrabouillmax) [inline app.html:12]: This can be deleted now?

File: app.html [hunk 2]
```diff
@@ -18,5 +26,9 @@
   <script src="graph_plot.js"></script>
   <script src="gradient_descent.js"></script>
   <script src="test_graph_plot.js"></script>
+  <div id="documentation" align="center">
+    <p style="color:black;font-size:50px;"> Gradient Descent Algorithm</p>
```
Comment (Scrabouillmax) [inline app.html:30]: Just a personal preference but I think it is better if at least the title is above the visualisation (the description can stay below). Your call :)

File: gradient_descent.js [hunk 3]
```diff
@@ -1,11 +1,36 @@
-class GradientDescent {
-  constructor(objective, h = 0.01, x_ini = [1,2], delta = 0.1) {
-    this.h = h;
+/** Mother class algorithm must do the following tasks
```
Comment (Scrabouillmax) [inline gradient_descent.js]:
```suggestion
/** Abstract class Algorithm is only used to define the following API for implemented algorithms:
```

Comment (Scrabouillmax) [reply gradient_descent.js]: In the future we could use typescript to be able to define the interfaces more precisely but it's good for now :)

File: gradient_descent.js [hunk 4]
```diff
@@ -43,17 +58,72 @@ class GradientDescent {
...
+    super(objective, x_ini, h, delta);
+    this.momentum = momentum;
+    this.currentgrad = x_ini.map(x => 0);
+  }
+
+  one_step() {
+    const gradient = this.differentiate(this.x);
+    this.currentgrad = this.currentgrad.map((e,i) => this.momentum * e + this.delta * gradient[i])
+    var norm = 0;
+    for (var i = 0, len = gradient.length; i < len; i++) {
```
Comment (Scrabouillmax) [inline gradient_descent.js]: Stop using var, use let instead 😂 
```suggestion
    for (let i = 0, len = gradient.length; i < len; i++) {
```

File: gradient_descent.js [hunk 5]
```diff
@@ -43,17 +58,72 @@ class GradientDescent {
...
+  constructor(objective, x_ini, h = 0.001, delta = 0.1, momentum = 0.9) {
+    super(objective, x_ini, h, delta);
+    this.momentum = momentum;
+    this.currentgrad = x_ini.map(x => 0);
+  }
+
+  one_step() {
+    const gradient = this.differentiate(this.x);
+    this.currentgrad = this.currentgrad.map((e,i) => this.momentum * e + this.delta * gradient[i])
+    var norm = 0;
```
Comment (Scrabouillmax) [inline gradient_descent.js]:
```suggestion
    let norm = 0;
```

File: test_graph_plot.js [hunk 6]
```diff
@@ -77,6 +101,19 @@ function test_square(h, eps, nlim) {
       .style("fill", "black")
   }
 
+  function onAlgorithmChanged(myalgorithm) {
```
Comment (Scrabouillmax) [inline test_graph_plot.js]:
```suggestion
  function onAlgorithmChanged(newAlgorithm) {
```

**Review by Scrabouillmax — APPROVED** [review]
Nice 🥇

File: app.html (hunk 1 above)
Comment (pierreosselin) [reply app.html:12]: Not really, it can be uncommented if we want to integrate latex language into <p> tags.

**Pull request merged**